File: src/main/java/tracker/BitcoinTracker.java
package tracker;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonElement;
import com.google.gson.JsonParser;
import java.io.BufferedInputStream;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.TreeMap;
import org.json.JSONObject;

public class BitcoinTracker {

  private String bitcoinEndpoint;
  private String currentPrice = "";
  private String currentTime = "";
  private String currency;
  private String startTime;
  private String endTime;
  private String jsonResult;

  public BitcoinTracker() {
    currency = "EUR";
    bitcoinEndpoint = "https://api.coindesk.com/v1/bpi/currentprice/" + currency + ".json";
  }

  public BitcoinTracker(String currency) {
    if (currency.length() < 3) {
      this.currency = "EUR";
    }
    this.currency = currency;
    bitcoinEndpoint = "https://api.coindesk.com/v1/bpi/currentprice/" + currency + ".json";
  }

  public BitcoinTracker(String startTime, String endTime) {
    this.startTime = startTime;
    this.endTime = endTime;
    bitcoinEndpoint =
        "https://api.coindesk.com/v1/bpi/historical/close.json" + "?start=" + this.startTime
            + "&end=" + this.endTime;
  }

  private void checkPrice() {
    HttpURLConnection urlConnection = null;
    StringBuilder result = new StringBuilder();

    try {
      URL url = new URL(bitcoinEndpoint);
      urlConnection = (HttpURLConnection) url.openConnection();
      int code = urlConnection.getResponseCode();

      if (code == 200) {
        try (InputStream in = new BufferedInputStream(urlConnection.getInputStream());
            BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(in))) {
          String line;
          while ((line = bufferedReader.readLine()) != null) {
            result.append(line);
          }
        }
      }
    } catch (IOException e) {
      e.printStackTrace();
    } finally {
      assert urlConnection != null;
      urlConnection.disconnect();
    }
    jsonResult = String.valueOf(result);

  }

  public String getCurrentPrice() {
    String result = parseResult();
    JSONObject obj = new JSONObject(result);

    currentPrice = obj.getJSONObject("bpi").getJSONObject(currency).getString("rate");
    currentTime = obj.getJSONObject("time").getString("updated");

    return currentPrice;
  }

  public TreeMap<String, String> getHistoricalPrice() {
    String result = parseResult();

    System.out.println(result);

    JSONObject obj = new JSONObject(result);
    String x = String.valueOf(obj.getJSONObject("bpi"))
        .replace("{", "")
        .replace("}", "")
        .replace("\"", "");

    TreeMap<String, String> historicalPrices = new TreeMap<>();
    String[] prices = x.split(",");
    for (String price : prices) {
      historicalPrices.put(price.split(":")[0], price.split(":")[1]);
    }

    return historicalPrices;
  }

  private String parseResult() {
    checkPrice();
    JsonParser jParser = new JsonParser();
    JsonElement jsonElement = jParser.parse(jsonResult);
    Gson gson = new GsonBuilder().setPrettyPrinting().disableHtmlEscaping().create();
    return gson.toJson(jsonElement);

  }

  public String getStartTime() {
    return startTime;
  }

  public String getEndTime() {
    return endTime;
  }

  public String getCurrency() {
    return currency;
  }

  public String getCurrentTime() {
    return currentTime;
  }

}

File: src/main/java/tracker/BitcoinTracker.kt
package tracker

import com.google.gson.*
import org.json.JSONObject
import java.io.*
import java.net.HttpURLConnection
import java.net.URL
import java.util.*

class BitcoinTracker {

    var bitcoinEndpoint:String = ""
    var marketPrice:String = ""
    var currentTime:String = ""
    var currency: String = ""
    var startTime: String = ""
    var endTime: String = ""
    var jsonResult: String = ""

    constructor()   {
        this.currency = "EUR"
        bitcoinEndpoint = "https://api.coindesk.com/v1/bpi/currentprice/" + this.currency +".json"
    }

    constructor(currency: String)   {
        if(currency.length < 3) {
            this.currency = "EUR"
        }
        this.currency = currency
        bitcoinEndpoint = "https://api.coindesk.com/v1/bpi/currentprice/" + this.currency +".json"
    }

    constructor(startTime: String,  endTime: String)    {
        this.startTime = startTime
        this.endTime = endTime
        bitcoinEndpoint = "https://api.coindesk.com/v1/bpi/historical/close.json" + "?start=" + this.startTime + "&end=" + this.endTime
    }

    private fun checkPrice()    {
        var urlConnection:HttpURLConnection? = null
        var result:StringBuilder = StringBuilder()

        try {
            var url = URL(bitcoinEndpoint)
            urlConnection = url.openConnection() as HttpURLConnection
            var codeResponse: Int = urlConnection.responseCode

            if (codeResponse == 200) {
                BufferedInputStream(urlConnection.inputStream).use { `in` ->
                    BufferedReader(InputStreamReader(`in`)).use { bufferedReader ->
                        result.append(bufferedReader.readLine())
                    }
                }
            }
        }   catch (e: IOException)   {
            e.printStackTrace()
        }   finally {
            urlConnection?.disconnect()
        }
        jsonResult =result.toString()
    }

    fun getCurrentPrice(): String   {
        val result = parseResponse()
        val obj = JSONObject(result)

        marketPrice = obj.getJSONObject("bpi").getJSONObject(currency).getString("rate")
        currentTime = obj.getJSONObject("time").getString("updated")

        return marketPrice
    }

    private fun parseResponse(): String {
        checkPrice()
        val jParser = JsonParser()
        val jsonElement = jParser.parse(jsonResult)
        val gSon:Gson = GsonBuilder().setPrettyPrinting().disableHtmlEscaping().create()
        return gSon.toJson(jsonElement)
    }

    fun getHistoricalPrice(): TreeMap<String, String> {
        val result = parseResponse()
        val obj = JSONObject(result)
        val x:String = obj.getJSONObject("bpi").toString()
            .replace("{", "")
            .replace("}", "")
            .replace("\"", "")

        val historicalPrices = TreeMap<String, String>()
        val prices = x.split(",".toRegex()).dropLastWhile({ it.isEmpty() }).toTypedArray()
        for (price in prices) {
            historicalPrices[price.split(":".toRegex()).dropLastWhile({ it.isEmpty() }).toTypedArray()[0]] = price.split(":".toRegex()).dropLastWhile({ it.isEmpty() }).toTypedArray()[1]
        }

        return historicalPrices
    }

}